File: WebApplication1/DrawBridge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using System.IO;
namespace RefrigtzW
{
    public class DrawBridge
    {
        //Iniatite Global Variable.
        public bool MovementsAStarGreedyHuristicFoundT = false;
        public bool IgnoreSelfObjectsT = false;
        public bool UsePenaltyRegardMechnisamT = true;
        public bool BestMovmentsT = false;
        public bool PredictHuristicT = true;
        public bool OnlySelfT = false;
        public bool AStarGreedyHuristicT = false;
        public bool ArrangmentsChanged = false;
        public static double MaxHuristicxB = -20000000000000000;
        public float Row, Column;
        public Color color;
        public ThinkingChess[] BridgeThinking = new ThinkingChess[AllDraw.BridgeMovments];
        public int[,] Table = null;
        public int Current = 0;
        public int Order;

        static void Log(Exception ex)
        {
            try
            {
                string stackTrace = ex.ToString();
                File.AppendAllText(FormRefrigtz.Root + "\\ErrorProgramRun.txt", stackTrace + ": On" + DateTime.Now.ToString()); // path of file where stack trace will be stored.
            }
            catch (Exception t) { Log(t); }
        }
        public bool MaxFound(ref bool MaxNotFound)
        {
            try
            {
                double a = ReturnHuristic();
                if (MaxHuristicxB < a)
                {
                    MaxNotFound = false;
                    if (ThinkingChess.MaxHuristicx < MaxHuristicxB)
                        ThinkingChess.MaxHuristicx = a;
                    MaxHuristicxB = a;
                    return true;
                }
            }
            catch (Exception t)
            {
                Log(t);

            }
            MaxNotFound = true;
            return false;
        }
        public double ReturnHuristic()
        {
            double a = 0;
            for (int ii = 0; ii < AllDraw.BridgeMovments; ii++)
                try
                {
                    a += BridgeThinking[ii].ReturnHuristic(-1, -1, Order);
                }
                catch (Exception t)
                {
                    Log(t);
                }

            return a;
        }


        //Constructor 1.
        public DrawBridge(bool MovementsAStarGreedyHuristicTFou, bool IgnoreSelfObject, bool UsePenaltyRegardMechnisa, bool BestMovment, bool PredictHurist, bool OnlySel, bool AStarGreedyHuris, bool Arrangments)
        {
            MovementsAStarGreedyHuristicFoundT = MovementsAStarGreedyHuristicTFou;
            IgnoreSelfObjectsT = IgnoreSelfObject;
            UsePenaltyRegardMechnisamT = UsePenaltyRegardMechnisa;
            BestMovmentsT = BestMovment;
            PredictHuristicT = PredictHurist;
            OnlySelfT = OnlySel;
            AStarGreedyHuristicT = AStarGreedyHuris;
            ArrangmentsChanged = Arrangments;
        }
        //constructor 2.
        public DrawBridge(bool MovementsAStarGreedyHuristicTFou, bool IgnoreSelfObject, bool UsePenaltyRegardMechnisa, bool BestMovment, bool PredictHurist, bool OnlySel, bool AStarGreedyHuris, bool Arrangments, float i, float j, Color a, int[,] Tab, int Ord, bool TB, int Cur//, ref FormRefrigtz THIS
            )
        {
            MovementsAStarGreedyHuristicFoundT = MovementsAStarGreedyHuristicTFou;
            IgnoreSelfObjectsT = IgnoreSelfObject;
            UsePenaltyRegardMechnisamT = UsePenaltyRegardMechnisa;
            BestMovmentsT = BestMovment;
            PredictHuristicT = PredictHurist;
            OnlySelfT = OnlySel;
            AStarGreedyHuristicT = AStarGreedyHuris;
            ArrangmentsChanged = Arrangments;
            //Initiate Global Variable By Local Parmenter.
            Table = new int[8, 8];
            for (int ii = 0; ii < 8; ii++)
                for (int jj = 0; jj < 8; jj++)
                    Table[ii, jj] = Tab[ii, jj];
            for (int ii = 0; ii < AllDraw.BridgeMovments; ii++)
                BridgeThinking[ii] = new ThinkingChess(MovementsAStarGreedyHuristicFoundT, IgnoreSelfObjectsT, UsePenaltyRegardMechnisamT, BestMovmentsT, PredictHuristicT, OnlySelfT, AStarGreedyHuristicT, ArrangmentsChanged, (int)i, (int)j, a, Tab, 16, Ord, TB, Cur, 4, 4);

            Row = i;
            Column = j;
            color = a;
            Order = Ord;
            Current = Cur;

        }
        //Clone a Copy.
        public void Clone(ref DrawBridge AA//, ref FormRefrigtz THIS
            )
        {
            int[,] Tab = new int[8, 8];
            for (int i = 0; i < 8; i++)
                for (int j = 0; j < 8; j++)
                    Tab[i, j] = this.Table[i, j];
            //Initiate a Constructed Brideges an Clone a Copy.
            AA = new DrawBridge(MovementsAStarGreedyHuristicFoundT, IgnoreSelfObjectsT, UsePenaltyRegardMechnisamT, BestMovmentsT, PredictHuristicT, OnlySelfT, AStarGreedyHuristicT, ArrangmentsChanged, this.Row, this.Column, this.color, this.Table, this.Order, false, this.Current);
            AA.ArrangmentsChanged = ArrangmentsChanged;
            for (int i = 0; i < AllDraw.BridgeMovments; i++)
            {
                try
                {
                    AA.BridgeThinking[i] = new ThinkingChess(MovementsAStarGreedyHuristicFoundT, IgnoreSelfObjectsT, UsePenaltyRegardMechnisamT, BestMovmentsT, PredictHuristicT, OnlySelfT, AStarGreedyHuristicT, ArrangmentsChanged, (int)this.Row, (int)this.Column);
                    this.BridgeThinking[i].Clone(ref AA.BridgeThinking[i]);
                }
                catch (Exception t)
                {
                    Log(t);
                    AA.BridgeThinking[i] = null;
                }
            }
            AA.Table = new int[8, 8];
            for (int ii = 0; ii < 8; ii++)
                for (int jj = 0; jj < 8; jj++)
                    AA.Table[ii, jj] = Tab[ii, jj];
            AA.Row = Row;
            AA.Column = Column;
            AA.Order = Order;
            AA.Current = Current;
            AA.color = color;

        }
        //Draw An Instatnt Brideges Images On the Table Method.
        public void DrawBridgeOnTable(ref Graphics g, int CellW, int CellH)
        {
            try
            {
                //Gray Color.
                if (color == Color.Gray)
                {
                    //Draw a Gray Bridges Instatnt Image On hte Tabe.
                    g.DrawImage(Image.FromFile(AllDraw.ImagesSubRoot + "BrG.png"), new Rectangle((int)(Row * (float)CellW), (int)(Column * (float)CellH), CellW, CellH));
                }
                else
                {
                    //Draw an Instatnt of Brown Bridges On the Table.
                    g.DrawImage(Image.FromFile(AllDraw.ImagesSubRoot + "BrB.png"), new Rectangle((int)(Row * (float)CellW), (int)(Column * (float)CellH), CellW, CellH));
                }
            }
            catch (Exception t)
            {
                Log(t);
            }
        }
    }
}
//End of Documents.

File: WebApplication1/DrawBridge.vb
Imports System.IO
Imports System.Drawing
Imports System.Text
Imports System.Linq
Imports System.Collections.Generic
Imports System
Namespace Refrigtz
    Public Class DrawBridge
        'Iniatite Global Variable.
        Public Row As Single, Column As Single
        Public color As Color
        Public BridgeThinking As ThinkingChess() = New ThinkingChess(AllDraw.BridgeMovments) {}
        Public Table As Integer(,) = Nothing
        Public Current As Integer = 0
        Public Order As Integer
        Shared Sub Log(ex As Exception)
            Try
                Dim stackTrace As String = ex.ToString()
                ' path of file where stack trace will be stored.
                File.AppendAllText("ErrorProgramRun.txt", stackTrace + ": On" + DateTime.Now.ToString())
            Catch t As Exception
                Log(t)
            End Try
        End Sub
        'Constructor 1.
        Public Sub New()
        End Sub
        'constructor 2.
        Public Sub New(i As Single, j As Single, a As Color, Tab As Integer(,), Ord As Integer, TB As Boolean,
            Cur As Integer)
            ' : base(i, j, a, Table)
            'Initiate Global Variable By Local Parmenter.
            Dim ii As Integer
            Table = Tab
            ii = 0
            While ii < AllDraw.BridgeMovments
                BridgeThinking(ii) = New ThinkingChess(CType(i, Integer), CType(j, Integer), a, Tab, 16, Ord,
                    TB, Cur, 4)
                ii += 1
            End While

            Row = i
            Column = j
            color = a
            Order = Ord

            Current = Cur
        End Sub
        'Clone a Copy.
        Public Sub Clone(ByRef AA As DrawBridge, ByRef THIS As FormRefrigtz)
            'Initiate a Constructed Brideges an Clone a Copy.
            AA = New DrawBridge(Me.Row, Me.Column, Me.color, Me.Table, Me.Order, False,
                Me.Current)
            Dim i As Integer
            i = 0
            While i < AllDraw.BridgeMovments
                Try
                    AA.BridgeThinking(i) = New ThinkingChess()
                    Me.BridgeThinking(i).Clone(AA.BridgeThinking(i), THIS)
                Catch t As Exception
                    Log(t)
                    AA.BridgeThinking(i) = Nothing
                End Try
                i += 1
            End While
        End Sub
        'Draw An Instatnt Brideges Images On the Table Method.
        Public Sub DrawBridgeOnTable(ByRef g As Graphics, CellW As Integer, CellH As Integer)
            Try
                'Gray Color.
                If color = Color.Gray Then
                    'Draw a Gray Bridges Instatnt Image On hte Tabe.
                    g.DrawImage(Image.FromFile(AllDraw.ImagesSubRoot + "BrG.png"), New Rectangle(CType((Row * CType(CellW, Single)), Integer), CType((Column * CType(CellH, Single)), Integer), CellW, CellH))
                Else
                    'Draw an Instatnt of Brown Bridges On the Table.
                    g.DrawImage(Image.FromFile(AllDraw.ImagesSubRoot + "BrB.png"), New Rectangle(CType((Row * CType(CellW, Single)), Integer), CType((Column * CType(CellH, Single)), Integer), CellW, CellH))
                End If
            Catch t As Exception
                Log(t)
            End Try
        End Sub
    End Class
End Namespace